### PR TITLE
[meshcop] add UDP proxy server

### PR DESCRIPTION
**Pull request by gjc13**

### CODE REVIEW

File: tests/meshcop/commissioner.cpp [hunk 1]
```diff
@@ -104,27 +105,50 @@ Commissioner::Commissioner(const uint8_t *     aPskcBin,
...
+int Commissioner::SetupProxyServer()
+{
+    int         ret;
+    int         optval = 1;
+    sockaddr_in addr;
+    addr.sin_family      = AF_INET;
+    addr.sin_port        = htons(kCommissionerProxyPort);
+    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+    mUdpListenFd         = socket(AF_INET, SOCK_DGRAM, 0);
+    setsockopt(mUdpListenFd, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval));
```
[the remark this answers is not in the document]
Comment (gjc13) [reply tests/meshcop/commissioner.cpp]: For debugging convenience.

File: src/common/udp_encapsulation_tlv.hpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+#ifndef OTBR_UDP_ENCAPSULATION_TLV_HPP_
+#define OTBR_UDP_ENCAPSULATION_TLV_HPP_
+
+#include "arpa/inet.h"
+#include "common/tlv.hpp"
+
+namespace ot {
+
+class UdpEncapsulationTlv : public Tlv
```
[the remark this answers is not in the document]
Comment (gjc13) [reply src/common/udp_encapsulation_tlv.hpp]: Updated

File: tests/meshcop/commissioner.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bukepo) [reply tests/meshcop/commissioner.cpp]: This could cause issues when the port is used by other instances or programs.

**Comment by gjc13** [comment]
@bukepo Added test code to meet code coverage requirement

File: src/commissioner/commissioner.cpp [hunk 3]
```diff
@@ -102,22 +103,29 @@ static int DummyKeyExport(void *               aContext,
...
-    addr.sin_port        = htons(kPortJoinerSession);
+    addr.sin_port        = htons(kJoinerSessionPort);
     mCoapAgent           = Coap::Agent::Create(SendCoap, this);
     mCoapToken           = rand();
     mCoapAgent->AddResource(mRelayReceiveHandler);
+    mCoapAgent->AddResource(mUdpRxHandler);
     mCommissionState = kStateInvalid;
     VerifyOrExit((mJoinerSessionClientFd = socket(AF_INET, SOCK_DGRAM, 0)) > 0);
     SuccessOrExit(connect(mJoinerSessionClientFd, reinterpret_cast<struct sockaddr *>(&addr), sizeof(addr)));
+    if (SetupProxyServer() != 0)
```
Comment (bukepo) [inline src/commissioner/commissioner.cpp]: We don't throw exceptions, so we should make the constructor always success. I suggest do such initialization in a separate method.

File: src/commissioner/commissioner.cpp [hunk 4]
```diff
@@ -128,10 +136,26 @@ void Commissioner::SetJoiner(const char *aPskdAscii, const SteeringData &aSteeri
     {
         delete mJoinerSession;
     }
-    mJoinerSession = new JoinerSession(kPortJoinerSession, aPskdAscii);
+    mJoinerSession = new JoinerSession(kJoinerSessionPort, aPskdAscii);
```
Comment (bukepo) [inline src/commissioner/commissioner.cpp]: check if `mJoinerSession` is `NULL`

File: .clang-format [hunk 5]
```diff
@@ -105,7 +105,7 @@ SpacesInContainerLiterals: true
 SpacesInCStyleCastParentheses: false
 SpacesInParentheses: false
 SpacesInSquareBrackets: false
-Standard:        Cpp11
+Standard:        Cpp03
```
Comment (bukepo) [inline .clang-format]: Shall we update this file in openthread/openthread?

File: src/commissioner/commissioner.cpp [hunk 6]
```diff
@@ -128,10 +136,26 @@ void Commissioner::SetJoiner(const char *aPskdAscii, const SteeringData &aSteeri
...
+int Commissioner::SetupProxyServer(void)
+{
+    int         ret;
+    int         optval = 1;
+    sockaddr_in addr;
+    addr.sin_family      = AF_INET;
+    addr.sin_port        = htons(kCommissionerProxyPort);
+    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+    mUdpListenFd         = socket(AF_INET, SOCK_DGRAM, 0);
+    setsockopt(mUdpListenFd, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval));
```
Comment (bukepo) [inline src/commissioner/commissioner.cpp]: Check return values.

File: src/commissioner/commissioner.cpp [hunk 7]
```diff
@@ -433,21 +459,34 @@ void Commissioner::Process(const fd_set &aReadFdSet, const fd_set &aWriteFdSet,
 
     if (FD_ISSET(mJoinerSessionClientFd, &aReadFdSet))
     {
-        struct sockaddr_in from_addr;
-        socklen_t          addrlen = sizeof(from_addr);
+        struct sockaddr_in fromAddr;
+        socklen_t          addrlen = sizeof(fromAddr);
+        ssize_t            n =
+            recvfrom(mJoinerSessionClientFd, buffer, sizeof(buffer), 0, (struct sockaddr *)(&fromAddr), &addrlen);
```
Comment (bukepo) [inline src/commissioner/commissioner.cpp]: c++ cast.

File: tests/meshcop/tmf_client.hpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * @file
+ *   The file is the header of tmf client and net topology data structure
+ */
+
+#ifndef OTBR_TMF_CLIENT_HPP_
+#define OTBR_TMF_CLIENT_HPP_
+
+#include <arpa/inet.h>
+#include <errno.h>
```
Comment (bukepo) [inline tests/meshcop/tmf_client.hpp]: not necessary in this header.

File: src/utils/misc.hpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
+/*
```
Comment (bukepo) [inline src/utils/misc.hpp]: duplicate with `commissioner_utils.hpp?

**Comment by gjc13** [comment]
@bukepo rebased to master

File: src/commissioner/commissioner.cpp [hunk 10]
```diff
@@ -431,23 +466,36 @@ void Commissioner::Process(const fd_set &aReadFdSet, const fd_set &aWriteFdSet,
...
             SendRelayTransmit(buffer, n);
         }
     }
+
+    if (mUdpListenFd > 0 && FD_ISSET(mUdpListenFd, &aReadFdSet))
+    {
+        struct sockaddr_in fromAddr;
+        socklen_t          addrlen = sizeof(fromAddr);
+
+        ssize_t n = recvfrom(mUdpListenFd, buffer, sizeof(buffer), 0, (struct sockaddr *)(&fromAddr), &addrlen);
```
Comment (bukepo) [inline src/commissioner/commissioner.cpp]: `static_cast`

File: src/commissioner/commissioner.cpp [hunk 11]
```diff
@@ -644,6 +692,74 @@ int Commissioner::SendRelayTransmit(uint8_t *aBuf, size_t aLength)
...
+                               void *                aContext)
+{
+    Commissioner * commissioner = static_cast<Commissioner *>(aContext);
+    int            tlvType;
+    uint16_t       length;
+    const uint8_t *payload  = aMessage.GetPayload(length);
+    uint16_t       destPort = 0;
+    sockaddr_in    addr;
+
+    if (commissioner->mUdpListenFd > 0)
```
Comment (bukepo) [inline src/commissioner/commissioner.cpp]: use `VerifyOrExit()` to reduce nested code?

File: src/commissioner/main.cpp [hunk 12]
```diff
@@ -60,16 +61,15 @@ int main(int argc, char **argv)
...
-    signal(SIGTERM, HandleSignal);
 
     srand(time(0));
 
     {
-        Commissioner commissioner(args.mPSKc, args.mKeepAliveInterval);
-        bool         joinerSetDone = false;
+        bool joinerSetDone = false;
 
+        srand(time(0));
```
Comment (bukepo) [inline src/commissioner/main.cpp]: duplicate with line 65?

File: src/commissioner/commissioner_constants.hpp [hunk 13]
```diff
@@ -76,6 +76,8 @@ enum
     kMbedDtlsHandshakeMaxTimeout = 60000, ///< dtls handshake min timeout
 
     kKEKSize = 32, ///< key encrypted key(KEK) size
+
+    kCommissionerProxyPort = 42873, ///< port commissioner use to proxy
```
Comment (bukepo) [inline src/commissioner/commissioner_constants.hpp]: make this a command line argument?

File: src/commissioner/main.cpp [hunk 14]
```diff
@@ -93,13 +99,20 @@ int main(int argc, char **argv)
...
             FD_ZERO(&errorFdSet);
+
             commissioner.UpdateFdSet(readFdSet, writeFdSet, errorFdSet, maxFd, timeout);
             rval = select(maxFd + 1, &readFdSet, &writeFdSet, &errorFdSet, &timeout);
             if (rval < 0)
             {
                 otbrLog(OTBR_LOG_ERR, "select() failed", strerror(errno));
                 break;
             }
+            else if (rval < 0 && errno == EINTR)
```
Comment (bukepo) [inline src/commissioner/main.cpp]: Will these lines be executed?

Comment (gjc13) [reply src/commissioner/main.cpp]: Updated

File: src/commissioner/commissioner_constants.hpp (hunk 13 above)
Comment (gjc13) [reply src/commissioner/commissioner_constants.hpp]: Updated

File: tests/meshcop/commissioner_proxy.hpp [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+#include "agent/coap.hpp"
+
+namespace ot {
+namespace BorderRouter {
+
+class CommissionerProxy
+{
+public:
+    /**
+     * The constructor to initialize Commissioner proxy
```
Comment (jwhui) [inline tests/meshcop/commissioner_proxy.hpp]: nit: use full sentences with period for doxygen briefs.

File: tests/meshcop/commissioner_proxy.hpp [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+     *
+     */
+    CommissionerProxy(int aProxyPort);
+
+    /**
+     * This method send data through commission proxy
+     *
+     * @param[in]   aDestAddr       Destination address to send data to
+     * @param[in]   aBuf            Buffer of data
+     * @param[in]   aLength         length to send
```
Comment (jwhui) [inline tests/meshcop/commissioner_proxy.hpp]: nit: be consistent with capitalizing first word of description

File: tests/meshcop/tmf_client.cpp [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+                            uint16_t       aLength,
+                            const uint8_t *aIp6,
+                            uint16_t       aPort,
+                            void *         aContext)
+{
+    TmfClient *client = static_cast<TmfClient *>(aContext);
+
+    return client->mProxy->SendTo(client->mDestAddr, aBuffer, aLength);
+
+    (void)aIp6;
```
Comment (jwhui) [inline tests/meshcop/tmf_client.cpp]: Put these in separate block above local variable declarations?

**Pull request closed without merging**